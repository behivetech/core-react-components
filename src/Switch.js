// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';

// Components
import MdcSwitch from '@material/react-switch';

// Styles
import '@material/react-switch/index.scss';
import './Switch.scss';

export default class Switch extends Component {
    constructor(props) {
        super(props);
        this.state = {
            checked: props.checked,
            indeterminate: false,
        };
        this.handleChange = this.handleChange.bind(this);
    }

    UNSAFE_componentWillReceiveProps({checked: nextChecked}) {
        if (this.state.checked !== nextChecked) {
            this.setState({checked: nextChecked});
        }
    }

    handleChange(event) {
        event.stopPropagation();
        this.setState({
            checked: event.target.checked,
            indeterminate: event.target.indeterminate,
        });
        this.props.onChange(event);
    }

    render() {
        const {
            disabled,
            id,
            name,
            value,
        } = this.props;

        return (
            <React.Fragment>
                <MdcSwitch
                    className="switch"
                    disabled={disabled}
                    id={id || name}
                    name={name || id}
                    onChange={this.handleChange}
                    value={value}
                />
            </React.Fragment>
        );
    }
}

Switch.propTypes = {
    /** ability to add additional className to the component for addistional styling from parent */
    className: PropTypes.string,
    /** Controls the if the checkbox is checked or not */
    checked: PropTypes.bool,
    /** sets the state of the checkbox to disabled */
    disabled: PropTypes.bool,
    /** The id attribute for the input of the checkbox */
    id: PropTypes.string,
    /** The name attribute for the input of the checkbox */
    name: PropTypes.string,
    /**
        Callback function for when the checkbox changes. It's recommended to use
        onClick for the changing event since IE doesn't play well with onChange
    */
    onChange: PropTypes.func,
    /**
        The values to be compared with the default value with will check or umncheck
        the checkbox. If it's an array, this will compare to see if the value exists
        in the checkbox.
    */
    value: PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
    ]),
};

Switch.defaultProps = {
    checked: false,
    onChange: () => null,
    value: true,
};
