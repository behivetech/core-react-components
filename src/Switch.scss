.switch {
    display: inline-flex;
    position: relative;
    margin: 0 10px 0;
}
