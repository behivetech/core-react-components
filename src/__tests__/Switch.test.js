import Switch from "../Switch";

describe('Switch', () => {
    const event = {
        stopPropagation: jest.fn(),
        target: {
            value: 'mock event value',
        },
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render with the correct props', () => {
        const wrapper = shallow(<Switch name="mockName" className="mock-class" />);

        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should render as checked', () => {
        const wrapper = shallow(<Switch name="mockName" checked />)

        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should render as disabled', () => {
        const wrapper = shallow(<Switch disabled name="mockName" checked />)

        expect(wrapper.debug()).toMatchSnapshot();
    });
});