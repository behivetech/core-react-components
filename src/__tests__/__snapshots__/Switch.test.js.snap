// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render as checked 1`] = `
<Fragment>
  <Switch
    checked={false}
    className="switch"
    disabled={false}
    id="mockName"
    name="mockName"
    onChange={[Function]}
    value={true}
  />
</Fragment>
`;

exports[`Switch should render as disabled 1`] = `
<Fragment>
  <Switch
    checked={false}
    className="switch"
    disabled={true}
    id="mockName"
    name="mockName"
    onChange={[Function]}
    value={true}
  />
</Fragment>
`;

exports[`Switch should render with a label 1`] = `
<Fragment>
  <Switch
    checked={false}
    className="switch"
    disabled={true}
    id="mockName"
    name="mockName"
    onChange={[Function]}
    value={true}
  />
</Fragment>
`;

exports[`Switch should render with the correct props 1`] = `
<Fragment>
  <Switch
    checked={false}
    className="switch"
    disabled={false}
    id="mockName"
    name="mockName"
    onChange={[Function]}
    value={true}
  />
</Fragment>
`;

exports[`Switch should set the name and id props correctly 1`] = `
<Fragment>
  <Switch
    checked={false}
    className="switch"
    disabled={false}
    id="mockId"
    name="mockName"
    onChange={[Function]}
    value={true}
  />
</Fragment>
`;

exports[`Switch should set the name and id props correctly 2`] = `
<Fragment>
  <Switch
    checked={false}
    className="switch"
    disabled={false}
    id="mockId"
    name="mockId"
    onChange={[Function]}
    value={true}
  />
</Fragment>
`;
